src/Compilers/Core/Portable/MetadataReference/MetadataReference.cs

```diff
@@ -301,50 +302,77 @@ internal static MetadataReference CreateFromAssemblyInternal(Assembly assembly)
             MetadataReferenceProperties properties,
             DocumentationProvider? documentation = null)
         {
+            // Note: returns MetadataReference and not PortableExecutableReference so that we can in future support assemblies that
+            // which are not backed by PE image.
+
             return CreateFromAssemblyInternal(assembly, properties, documentation);
         }
 
         internal static PortableExecutableReference CreateFromAssemblyInternal(
             Assembly assembly,
             MetadataReferenceProperties properties,
-            DocumentationProvider? documentation = null)
-        {
-            // Note: returns MetadataReference and not PortableExecutableReference so that we can in future support assemblies that
-            // which are not backed by PE image.
+            DocumentationProvider? documentation = null) =>
+            CreateFromAssemblyOrNullInternal(assembly, properties, documentation, throwOnFailure: true)!;
 
+        internal static unsafe PortableExecutableReference? CreateFromAssemblyOrNullInternal(
+            Assembly assembly,
+            MetadataReferenceProperties properties,
+            DocumentationProvider? documentation = null,
+            bool throwOnFailure = false)
```

Comment:
I'm not sure I see the reasoning for this parameter. What prompted this part of the change, particularly since this seems to be a behavior change for the default case of calling the method?